src/rabbitmq.js

```diff
@@ -0,0 +1,354 @@
+          .catch((err) => {
+            this.log.error({ err: err }, 'an error occured creating channel')
+            throw err
+          })
+      })
+      .then((channel) => {
+        this.log.info('created channel')
+        this.channel = channel
+        this.channel.on('error', this._channelErrorHandler.bind(this))
+      })
+  }
+
+  /**
+   * Subscribe to a specific direct queue.
+   *
+   * @param {String} queue Queue name.
+   * @param {Function} handler Handler for jobs.
+   * @return {Promise} Promise that is resolved once queue is subscribed.
+   */
+  subscribeToQueue (queue: string, handler: Function): Promise {
+    const log = this.log.child({
+      method: 'subscribeToQueue',
+      queue: queue
+    })
+    log.info('subscribing to queue')
+    if (!this._isConnected()) {
+      return Promise.reject(new Error('you must .connect() before subscribing'))
+    }
+    if (!isFunction(handler)) {
+      log.error('handler must be a function')
+      return Promise.reject(
+        new Error(`handler for ${queue} must be a function`)
+      )
+    }
+    if (this.subscribed.has(`queue:::${queue}`)) {
+      log.warn('already subscribed to queue')
+      return Promise.resolve()
+    }
+    return Promise
+      .resolve(this.channel.assertQueue(queue, { durable: true }))
+      .then(() => {
+        log.info('queue asserted, binding queue')
+        this.subscriptions = this.subscriptions.set(queue, handler)
+        this.subscribed = this.subscribed.add(`queue:::${queue}`)
+      })
+  }
+
+  /**
+   * Subcribe to fanout exchange.
+   *
+   * @param {String} exchange Name of fanout exchange.
+   * @param {Function} handler Handler for jobs.
+   * @return {Promise} Promise resolved once subscribed.
+   */
+  subscribeToFanoutExchange (exchange: string, handler: Function): Promise {
+    return this._subscribeToExchange({
+      exchange: exchange,
+      type: 'fanout',
+      handler: handler
+    })
+  }
+
+  /**
+   * Subscribe to topic exchange.
+   *
+   * @param {String} exchange Name of topic exchange.
+   * @param {String} routingKey Routing key for topic exchange.
+   * @param {Function} handler Handler for jobs.
+   * @return {Promise} Promise resolved once subscribed.
+   */
+  subscribeToTopicExchange (
+    exchange: string,
+    routingKey: string,
+    handler: Function
+  ): Promise {
+    return this._subscribeToExchange({
+      exchange: exchange,
+      type: 'topic',
+      routingKey: routingKey,
+      handler: handler
+    })
+  }
+
+  /**
+   * Start consuming from subscribed queues.
+   *
+   * @return {Promise} Promise resolved when all queues consuming.
+   */
+  consume (): Promise {
+    const log = this.log.child({ method: 'consume' })
+    log.info('starting to consume')
+    if (!this._isConnected()) {
+      return Promise.reject(new Error('you must .connect() before consuming'))
+    }
+    const subscriptions = this.subscriptions
+    this.subscriptions = new Immutable.Map()
```

Comment:
the line before I copy the reference to the Map to the local scope, then replace the one that exists on the model.

> I do this because I want to deal with the subscriptions that I have, and not worry about the object changing under me